src/fairseq2/recipes/lm/_online_finetune/_remote_hf.py

```diff
@@ -0,0 +1,328 @@
+            padding="longest",
+            truncation=True,
+        )
+
+    def postprocess(self, model_outputs, function_to_apply=None, top_k=1, _legacy=True):
+        return model_outputs["scores"].cpu().float().item()
+
+
+class RemoteHFModel:
+    def __init__(
+        self,
+        ray_actor_name: str,
+        num_replicas: int,
+        hf_engine_args: HFEngineArgs,
+        sampling_params: dict,
+        init_update_process_group: bool,
+        gangs: Gangs,
+    ):
+        self._gangs = gangs
+
+        self.num_replicas = num_replicas
+        self.ray_actor_name = ray_actor_name
+
+        self.hf_workers = []
+        self.update_process_groups = []
+
+        if gangs.dp.rank != 0 and gangs.tp.rank != 0:
+            raise ValueError("hf worker should only be initialized on DP & TP rank 0")
+
+        for replica_i in range(self.num_replicas):
+            self.setup_replica(replica_i, hf_engine_args, init_update_process_group)
+            # gangs.root.barrier()/
+
+        # populate sampling params using all values that were passed in the config
+        self.sampling_params = SamplingParams(**sampling_params)
+
+        self._hf_engine_args = hf_engine_args
+
+    def setup_replica(self, replica_i: int, hf_engine_args, init_update_process_group):
+        if (
+            len(self.hf_workers) != replica_i
+            or len(self.update_process_groups) != replica_i
+        ):
+            raise RuntimeError(
+                "new replica is being created while previous ones are not setup yet"
+            )
+
+        hf_worker = self.setup_hf_worker(
+            f"{self.ray_actor_name}_{replica_i}", hf_engine_args, self._gangs
+        )
+        self.hf_workers.append(hf_worker)
+
+        update_process_group = self.setup_process_group_for_model_sync(
+            replica_i, hf_engine_args.tensor_parallel_size, init_update_process_group
+        )
+        self.update_process_groups.append(update_process_group)
+
+        log.info(f"Replica {replica_i} setup completed")
+
+    def setup_hf_worker(
+        self, ray_actor_name, hf_engine_args: HFEngineArgs, gangs: Gangs
+    ):
+
+        pg_inference = placement_group(
+            [{"GPU": 1, "CPU": 0}] * hf_engine_args.tensor_parallel_size,
+            strategy="STRICT_PACK",
+        )
+
+        ray.get(pg_inference.ready())
+
+        scheduling_inference = PlacementGroupSchedulingStrategy(
+            placement_group=pg_inference,
+            placement_group_capture_child_tasks=True,
+            placement_group_bundle_index=0,
+        )
+
+        llm = NoEnvPipeline.options(
+            name=ray_actor_name,
+            num_cpus=0,
+            num_gpus=0,
+            scheduling_strategy=scheduling_inference,
+            get_if_exists=True,
+        ).remote()
+
+        # we block here until the engine is initialized
+        ray.get(llm.is_ready.remote())
+
+        return llm
+
+    def setup_process_group_for_model_sync(
```

Comment:
i don't know if we need to replicate all of the _remote_vllm.py functions like this